### PR TITLE
Unexpected result of `getSelectText()`

### PR DESCRIPTION
**Pull request by AgatZan**
What I mean by "unexpected".

neovide_noybscCG6q

Maybe it's just because I don't really understand the use case of `getSelectedText()`, but I guess that function will catch the text **now** under selection, but the `'<` mark contain **last** selected region.

### CODE REVIEW

**Comment by AgatZan** [comment]
Bug, when `:'<,'>Scratch` `vim.api.nvim_get_mode().mode=="n"`

**Comment by AgatZan** [comment]
> Bug, when `:'<,'>Scratch`
> 
> ```lua
> vim.api.nvim_get_mode().mode == "n"
> ```

Also, find 1 cause. It is obvious (especially for me) that when using keybinds, `args.range == 0`, so either make something with keymaps or specify it in the user command. I choose 2.

And another problem is that the use of `:'<,'>Scratch` also fails to select because it is humorous, but the visual selection at call time is not visually represented **AS-IS**, and the saved cursor position at the beginning of the selection point is not correct. I have fixed this problem.

**Comment by AgatZan** [comment]
Old realisation works correctly when `:'<,'>` and doesn't work on mappings. 
New realisation works on mappings correctly and sometimes doesn't on `:'<,'>`.

File: tests/test_select.lua [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+--   )
+-- end
+T["param"]["new"]["workd"] = function(selection_mode, coord)
+  select_wise(coord, selection_mode)
+  MiniTest.expect.equality(
+    table_select(BUFFER_TEXT, coord, selection_mode),
+    child.lua_func(getSelectedText, ".", selection_mode)
+  )
+end
+-- T["param"]["old"]["at_command"] = function(selection_mode, coord)
```
Comment (AgatZan) [inline tests/test_select.lua:125]: Sometimes work, sometimes it doesn't, but I replace it so it doesn't really matter.

**Comment by LintaoAmons** [comment]
Hi, this looks really good!

I could merge this PR right now~ But there is a little problem I met when I was testing it locally (`:1,5Scratch`).

Here is the error msg and some info might be helpful. But if you think it not a big issue, I also think there's no need to bother with it right now.(Cause I don't think many user will user this command this way)


Error executing Lua callback: ...uments/oatnil/release/scratch.nvim/lua/scratch/utils.lua:89: attempt to index a nil value
stack traceback:
	...uments/oatnil/release/scratch.nvim/lua/scratch/utils.lua:89: in function 'getSelectedText'
	...ts/oatnil/release/scratch.nvim/plugin/scratch_plugin.lua:24: in function <...ts/oatnil/release/scratch.nvim/plugin/scratch_plugin.lua:18>


## CodeCompanion ─

Looking at the error and the code, the issue is in the `getSelectedText` function where it's trying to access `text[1]` but `text` might be empty or nil. Let's add some error checking to make the function more robust.

Here's the fixed version of the `getSelectedText` function:

```lua
local function getSelectedText(mark, selection_mode)
    local pos1 = vim.fn.getpos("v")
    local pos2 = vim.fn.getpos(mark)
    local lines = {}
    
    -- Validate positions
    if not pos1 or not pos2 then
        return lines
    end
    
    local start_row, start_col = pos1[2], pos1[3]
    local end_row, end_col = pos2[2], pos2[3]
    
    -- Get text from buffer
    local text = vim.api.nvim_buf_get_lines(0, start_row - 1, end_row, true)
    
    -- Check if text is empty
    if not text or #text == 0 then
        return lines
    end
    
    end_row = end_row - start_row + 1
    start_row = 1
    
    if selection_mode == "v" then
        if text[1] then
            table.insert(lines, text[1]:sub(start_col))
            for i = start_row + 1, end_row do
                if text[i] then
                    table.insert(lines, text[i])
                end
            end
            if lines[end_row] then
                lines[end_row] = lines[end_row]:sub(1, end_col)
            end
        end
    elseif selection_mode == "V" then
        for i = start_row, end_row do
            if text[i] then
                table.insert(lines, text[i])
            end
        end
    elseif selection_mode == vim.api.nvim_replace_termcodes("<C-V>", true, true, true) then
        for i = start_row, end_row do
            if text[i] then
                table.insert(lines, text[i]:sub(start_col, end_col))
            end
        end
    end
    
    return lines
end
```

**Comment by AgatZan** [comment]
@LintaoAmons Reason of these issue because i'm not reacting correctly in `:Scratch`. 

There are two use-flow branches: select and range. When I process a range of branches, I think of it as selection, and it is a bug.

```lua
  if mode ~= "n" then
    opts = { content = utils.getSelectedText(".", mode) }
  elseif args.range > 0 then
    opts = { content = vim.api.nvim_buf_get_lines(0, args.line1, args.line2, true) }
  end
```
I guess that correct way. But now you must use keybinding to perform select branch. Maybe there is a magical way to determine in command mode whether it is selected or just an range, but I don't know of that way.

**Comment by LintaoAmons** [comment]
I saw you pushed some new commits, let me know if you think it's ready to merge

**Comment by AgatZan** [comment]
@LintaoAmons Specifically within the work of this pull request, I wanted to fix `getSelectText`, I fixed it, and made some set of tests that would verify something.

Generally speaking, it is necessary to write tests for the `plugin/scratch_plugin.lua` file, but I have no idea whether they are required within this pull request or whether they should be implemented separately.

**Comment by LintaoAmons** [comment]
I would say it's not required within this PR. I tested my workflow and it's fine. So if you are confident (comfortable) I could merge it anytime

**Comment by AgatZan** [comment]
And how to understand it.

The mode is determined correctly.
If call `getSelectedText` with a api-call to get mode, it gives the correct result
```
  Failed expectation for equality.
  Left:  { "some", "text", "inse", "now" }
  Right: { "some", "text here will be", "inserted", "now" }
  Traceback:
    tests/test_plugin.lua:112
NOTE in tests/test_plugin.lua | Scratch | select_branch | parameter | some_text + args { "\22", { 1, 1, 4, 4 } }: |1>{
  blocking = false,
  mode = "\22"
}
NOTE in tests/test_plugin.lua | Scratch | select_branch | parameter | some_text + args { "\22", { 1, 1, 4, 4 } }: |2>{ "some", "text", "inse", "now" }
```
but if you do it using keymap, it does not give the correct result

**Comment by AgatZan** [comment]
@LintaoAmons Can you please stop [this](https://github.com/LintaoAmons/scratch.nvim/actions/runs/11757073491) action, because i can't and it in deadlock.

**Comment by LintaoAmons** [comment]
Hi @AgatZan , I saw you haven't pushed any new commits now, 
Should I merge it?

I tested it locally with my workflow. It's working fine.
I haven't spent time getting into the tests, so it's fine to have those tests failing now.

**Pull request merged**